.github/workflows/ci-ubuntu.yml

```diff
@@ -118,38 +89,75 @@ jobs:
           rm -rf agda
 
       - name: Download and install fix-whitespace
-        if: steps.cache-cabal.outputs.cache-hit != 'true'
+        if: steps.cache-external-restore.outputs.cache-hit != 'true'
         run: |
           git clone https://github.com/agda/fix-whitespace --depth=1
           cd fix-whitespace
           ${{ env.CABAL_INSTALL }} fix-whitespace.cabal
           cd ..
 
+      - name: Save external dependencies cache
+        if: steps.cache-external-restore.outputs.cache-hit != 'true'
+        uses: actions/cache/save@v3
+        id: cache-external-save
+        with:
+          path: ${{ env.CACHE_PATHS }}
+          key: ${{ steps.cache-external-restore.outputs.cache-primary-key }}
+
+      ########################################################################
+      ## CHECKOUT
+      ########################################################################
+
+      # By default github actions do not pull the repo
+      - name: Checkout cubical
+        uses: actions/checkout@v3
+
       ########################################################################
       ## TESTING
       ########################################################################
 
+      - name: Restore library cache
+        uses: actions/cache/restore@v3
+        id: cache-library-restore
+        with:
+          path: ./_build
+          key: library-${{ runner.os }}-${{ env.GHC_VERSION }}-${{ env.CABAL_VERSION }}-${{ env.AGDA_BRANCH }}-${{ hashFiles('Cubical/**') }}
+          restore-keys: |
+            library-${{ runner.os }}-${{ env.GHC_VERSION }}-${{ env.CABAL_VERSION }}-${{ env.AGDA_BRANCH }}-
+
+      - name: Put cabal programs in PATH
+        run: echo "~/.cabal/bin" >> $GITHUB_PATH
+
       - name: Test cubical
+        if: steps.cache-library-restore.outputs.cache-hit != 'true'
```

Comment:
This is a bit of an unconventional setup, I'd say, to skip the main step of a job.
Currently only changes under `Cubical/` will trigger this step, judging from the cache key.
What about changes in `cubical.agda-lib`, or `Everythings.hs` in the root directory?
What about new files added to the root?

Could you say a bit more about scenarios where it would be the right thing to skip the main step?

Reply:
I've miseed this case because I've been mainly working on files under `Cubical/` and considered everything else more or less fixed, my apologies!

Skipping any step is desirable if it would be a recomputation of already known data. If it's correctly implemented of course :-)

Reply:
@andreasabel
Adding new files in the root directory will incur additional costs as you'll need to also modifiy the CI config. Not caching will waste 50 minutes of compute on every re-run. Human time may be more valuable than machine time in this case, but changing project structure requires human attention anyway.
What's your decision on this?

Reply:
@cmcmA20: I was hoping for a more concrete answer to my question, like enumerating some concrete scenarios, so that one can better judge what is the right approach here:

> Could you say a bit more about scenarios when it would be the right thing to skip the main step?

Reply:
@andreasabel:
Forgive my opacity, I think skipping the main step on cache hit is natural and right thing to do. Here I assume that building project on CI is equivalent to building it locally with `make`. If both Agda and project files are unchanged, and you haven't explicitly dropped the cache (like running `make clean`), rebuilding/redeploying anything should be considered a bug in the workflow or the build system.

Maybe I don't see something obvious, please tell me where's the error in reasoning.

Reply:
Thanks for your reply.  I am not convinced yet:

You currently have in the cache key `hashFiles('Cubical/**', 'cubical.agda-lib', 'Everythings.hs')`.
So if someone changes the makefile and none of these other files and opens a PR, CI would skip the step one wants to test (provided caches have not been deleted).

I think If you insist that the main step be skipped if nothing changed, you have to define the _nothing_ correctly!  Also, it add some subtlety, and needs some extra maintenance (who is going to do this?).
So, I would advise against it.

>  I think skipping the main step on cache hit is natural and right thing to do.

I would prefer to also hear a concrete scenario.  
The scenario I can think of is when a PR just changes a `.md` file like `CONTRIBUTING`, or the `.gitignore` file.  But aren't these cases rare enough to ignore them?
Do you have other scenarios?

Reply:
Thank you, Andreas, for uncovering the weak points, I really appreciate it.
This design provides modest benefits, but the effort to get it right seems unreasonably high.

>CONTRIBUTING, or the .gitignore file. But aren't these cases rare enough to ignore them? Do you have other scenarios?
I had no scenarios in mind besides documentation changes.